docs/enhancements/0001-worker-pods.md

```diff
@@ -0,0 +1,233 @@
+                                │                       │           │                 │
+                                │                       │           │ NodeModuleState │
+┌────────────┐                  │ ┌─────────────────┐   │           │                 │
+│            │                  │ │                 │   │           └────────▲────────┘
+│ KMM worker ◄──────────────────┼─┤ NodeModuleState │   │                    │
+│    Pod     │  creates & owns  │ │ controller      ├───┼────────────────────┘
+│            │                  │ │                 │   │   watches &
+└──────┬─────┘                  │ └─────────────────┘   │   updates status
+       │                        │                       │
+       │                        │          KMM operator │
+       │                        └───────────────────────┘
+       │
+       │
+       │                        ┌──────────────────────┐
+       │                        │                      │
+       └────────────────────────► registry.example.org │
+                pulls from      │                      │
+                                └──────────────────────┘
+```
+
+#### The operator
+
+The operator runs two main controllers: the `Module` controller and the `NodeModuleState` controller.
+
+The `Module` controller watches both `Module` and `Node` resources to determine which kmods should be loaded on which
+`Node`.
+It creates one `NodeModuleState` resource per `Node` and `Module` that targets that node.  
+`NodeModuleState` resources represent the desired state of a `Module` on a node; they contain all the required data to
+pull the kmod image and load the module properly.  
+The `Module` controller watches updates to the `status` of `NodeModuleState` objects and aggregates that data into the
+`Module` `status`.
+
+The `NodeModuleState` controller watches `NodeModuleState` resources and determines when a worker Pod needs to be
+started on a node by looking at the difference between the desired state (spec) and the current one (status).
+When the Pod has completed, the controller reads its outcome and updates the `NodeModuleState` status accordingly.
+
+When updating the `NodeModuleState` status, the controller includes a timestamp corresponding to the last time a worker
+Pod ran on the node.
+This timestamp can be used to verify if a new worker Pod needs to run when the KMM operator restarts; if it is older
+than the last transition of the `Node`'s `Ready` condition, then the node may have rebooted and we potentially need to
+load the kernel modules.
+
+#### The worker Pods
+
+Worker Pods are created by the `NodeModuleState` controller.
+They run KMM code that reads the configuration file, pulls and extracts kmod OCI images and uses `modprobe` to load or
+unload kmods.  
+The configuration file is provided as an annotation to the Pod object and mounted in its filesystem via the [Downward
```

Comment:
Why did we chose this approach instead of passing cmd args or env variables to the worker pod?
Seems like downard API is used in order to expose to the pod information about himself such as its name, namespace etc.

Here we are injecting a configuration file to the pod's metadata to then use the downward API. Isn't it sort of abusing it?

Furthermore, if we insist on a configuration file, then why we don't use the configmap for that? (even if we read the output from the pod instead of the configmap).

Reply:
> Here we are injecting a configuration file to the pod's metadata to then use the downward API. Isn't it sort of abusing it?

I don't see how it constitutes an abuse. Using an annotation makes the config easily readable when getting the Pod with `kubectl`, as opposed to many environment variables or CLI flags.